### PR TITLE
Fix avatar enumable

### LINKED ISSUE #718: Uploaded custom avatars enumerable and downloadable w/o login despite "Require Sign In View" being set
- Gitea version (or commit ref): 1.0.1
- Operating system: Debian 8.7
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [] Yes (provide example URL) BUT
  - [ ] No
  - [x] Not relevant as the public instance doesn't have the setting in question set
- Log gist: https://gist.github.com/2ion/01f673323560ebe841dc93696ce2fd96

## Description

Uploaded custom avatars remain enumerable by their numerical index without having to log in despite the setting "require sign in view" being set. This leaks information about local users (reverse image search, "real" profile pictures, etc)  which is probably not intended. I suggest to require a login in this situtation.

```bash
#!/bin/bash
BASEURL=$1
BOUND=10000
check_http_status()
{
    local code=$(curl -o/dev/null --silent --head --write-out '%{http_code}' "$1")
    if (( code == 200 )) ; then
        return 0
    else
        return 1
    fi
}
for ((i=1;i<$BOUND;i++)); do
    URL="$BASEURL/avatars/$i"
    if check_http_status "$URL" ; then
        (wget -qO "$i" "$URL" &)
        echo -n +
    else
        echo -n _
    fi
done
echo
```
Use as:
```
./thescript https://your.private.gitea.instance.sexy
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/41210128-uploaded-custom-avatars-enumerable-and-downloadable-w-o-login-despite-require-sign-in-view-being-set?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by lunny**
fix #718

### CODE REVIEW

File: models/migrations/migrations.go [hunk 1]
```diff
@@ -88,6 +88,8 @@ var migrations = []Migration{
 	NewMigration("add external login user", addExternalLoginUser),
 	// v19 -> v20
 	NewMigration("generate and migrate Git hooks", generateAndMigrateGitHooks),
+	// v20 -> v21
+	NewMigration("use new avtar path name for security reason", useNewNameAvatars),
```
Comment (strk) [inline models/migrations/migrations.go:92]: Typo

**Comment by strk** [comment]
LGTM

**Review by appleboy — APPROVED** [review]

**Comment by andreynering** [comment]
LGTM, but I tried in a dev environment with a user that didn't have an avatar, and it generated a new one. Is this the desired behavior?

**Comment by lunny** [comment]
No. I think that's not a desired behavior. The migration will rename all the avatars on the data/avatars from <id> to <md5> and the upload avatar name will be <md5> not <id>. Only these two changes.

**Comment by appleboy** [comment]
LGTM

**Comment by gsantner** [comment]
Pulled to my installation. Now user avatars are not loaded, and wrong org ones.
Somehow for organizations the same icons get always loaded (there should be different ones).  
Did a restart, and gitea has write access to avatar dir.

grafik

**Comment by lunny** [comment]
Please fire an issue.

**Comment by richmahn** [comment]
This PR is making all my custom avatars (I have 2890 users, many which have custom avatars uploaded) be the same avatar! I added a few Printfs to the Migration, migrating up from v15, and I see this:

```Renaming /home/git/gogs-avatars/4613 to /home/git/gogs-avatars/f295aae8bd4b6dca44ad1c35bcf99ad7
USER TO UPDATE: 4613, {f295aae8bd4b6dca44ad1c35bcf99ad7 true}
2017/05/07 11:07:51 [I] [SQL] UPDATE `user` SET `avatar` = ?  [f295aae8bd4b6dca44ad1c35bcf99ad7]
```

As you can see, there is no WHERE condition to set id = 4613!

**Comment by lunny** [comment]
@richmahn what's your database?

**Pull request merged**